054-infinite-auto-restart-of-Kafka-connectors.md

```diff
@@ -0,0 +1,160 @@
+
+## Current situation
+
+Strimzi 0.33.0 introduced support for automatically restarting connectors in Apache Kafka Connect clusters (when managed using the `KafkaConnector` resources). This feature is disabled by default, but can be enabled in the `KafkaConnectors` `.spec` section:
+
+```yaml
+apiVersion: strimzi.io/v1beta2
+kind: KafkaConnector
+metadata:
+  name: my-source-connector
+spec:
+  # ...
+  autoRestart:
+    enabled: false
+  # ...
+```
+
+When enabled, it works like this:
+* The connector is restarted up to 7 times
+* The times of the auto-restarts are:
+    * Immediate
+    * 2+ minutes after previous restart
+    * 6+ minutes after previous restart
+    * 12+ minutes …
+    * 20+ minutes …
+    * 30+ minutes …
+    * 42+ minutes …
+    * _There is no exact time, so it will be restarted when the next reconciliation happens and the time condition is fulfilled (i.e. restart after 6+ minutes can happen for example after 7 minutes in the next suitable periodical reconciliation)._
+* After 7th restart, the connector will not be automatically restarted anymore
+* When the container runs successfully for at least the backoff interval for the restart counter to be reset.
+  E.g. if it was restarted 4 times, it needs to run for 20 minutes for the restart counter to be reset to 0 and the restart sequence start from the beginning in case of the next failure.
+
+## Motivation
+
+The auto-restart functionality works fine. But in some cases, it would be useful to have more flexibility including a possibility to retry the restarts indefinitely.
+Infinite restarts allow the auto-restarting to be used for even longer outages which for example take the whole weekend and it increases the value of this feature.
+
+## Proposal
+
+A new field `maxRestarts` will be added to the `autoRestart` section of the `KafkaConnector` CRD:
+
+```yaml
+apiVersion: strimzi.io/v1beta2
+kind: KafkaConnector
+metadata:
+  name: my-source-connector
+spec:
+  # ...
+  autoRestart:
+    enabled: false
+    maxRestarts: 20
+  # ...
+```
+
+The `maxRestarts` field will default to null (not set).
+And when it is not set (set to null), the operator will attempt to restart the connector infinitely.
+When a user sets `maxRestarts` to a specific value, the operator will attempt to restart the operator only for a given number of attempts and then give up on it.
+In that case, the user would need to restart it later manually.
+
+**The solution proposed by this proposal is not fully backwards compatible.**
+**Please read the _Compatibility_ section for more details.**
+
+The current timing will be unchanged.
+It will be calculated based on the following formula where the `restartCount` means the number of restarts that have already happened.
```

Comment:
What happens to the restartCount if the operator crashes and it is restarted? Is the restartCount stored somehow (annotation?) in the KafkaConnector resource or it just restarts from 0?

Reply:
There is no change planned to how it is implemented today. It is stored in the status of the custom resource.